### ISSUE TITLE
Document Manager Enhancement Discussion

### ISSUE THREAD

**Issue by JD-Howard** [Autodesk-AutoCAD/AutoLispExt#30]
**Is your feature request related to a problem? Please describe.**
Many of the intelligence features I want in the AutoLispExt require foraging for information across PRJ, Workspace and even on fragments of LSP files. I've already started submitting PR's to help with that, but this is where I want to do some preliminary discussion on how it will be organized and what it will provide.

**Describe the solution you'd like**
This will most likely be a class so that getters can be used to dynamically serve the best available content. This will likely be a single pre-initialized instance called DocumentManager and that document manager will probably be the only export. The constructor of the document manager is intended to create and manage all the event hooks such as:
- vscode.workspace.onDidOpenTextDocument
- vscode.workspace.onDidCloseTextDocument
- vscode.workspace.onDidRenameFiles
- vscode.workspace.onDidChangeWorkspaceFolders (has to reset the FileSystemWatcher events)
- vscode.FileSystemWatcher.onDidDelete
- vscode.FileSystemWatcher.onDidCreate
- vscode.FileSystemWatcher.onDidChange (not real sure this one will be necessary)

**Note:** The recent PR to natively add ReadonlyDocument's to the LspFileNode make it so that the PRJ events do not need to inform the DocumentManager about anything that happened in that space.

Another thing this will do is scan the workspace and find all the LSP files creating ReadonlyDocuments for each of them that get stored in a Map using their fspath as the key; this is to support multiple files with the same name. This global repo will take 3rd precedence for the content delivery system. Second precedence will be given to the PRJ referenced files and top priority will be given to opened documents; when aware of them that is. Still have a technical debt to figure out which files were already opened in the workspace reload when VSCode opened. However, once any of those file tabs get activated the above events will let us know and they will begin taking top priority.

There is performance concerns about reading the contents of all these files when our extension is first loading. It does not lock the interface, but it does in fact cause a _scanning_ bar to show and some features aren't 100% available until it completes. It doesn't take long at all, even with 10mb of LSP files in your workspace, but a bit unsightly. Because of this, I am wondering if we should make a way to create a ReadonlyDocument without the actual (content) initialize that could demand load when the user is actually doing something that tries to use that data. Another option is to create the DM Map entries with null and let the document manager demand load them when requested.

**Additional context**
The expected common access point for this will be through the ContextManager.Documents getter.

**Comment by JD-Howard** [comment]
Proposed including of a general purpose document selector detection. This could be used to eliminate a lot of inference around the code base and would directly help the DocumentManager determine if an _opened_ document is valid to be included in its result queries.
```typescript
getSelectorType(fspath: string): string { return DocumentManager.getSelectorType(fspath); }
public static getSelectorType(fspath: string): string {
	const ext: string = fspath.toUpperCase().slice(-4);
	switch (ext) {
		case ".LSP": return "autolisp";
		case ".MNL": return "autolisp";
		case ".PRJ": return "autolispprj";
		case ".DCL": return "autolispdcl";
		default: return "";
	}
}
```

**Comment by LanDong-adsk** [comment]
released in v 1.4.0
